### ISSUE TITLE
Add option not to version file or directory moving/renaming

### ISSUE THREAD

**Issue by santanoce** [syncthing/syncthing#8446]
Hello, I've been using Syncthing for years and I work in an environment in which I usually rename or move files and folders; unfortunately this results in huge sizes of the .stversions folder, because every renaming or moving is considered a change in the file/directory. I've read a lot of posts about this (some date back to 2015!) and a lot of people have my same problem. I totally understand that this is expected behavior, but I'm asking if there's the possibility to add an option in the file versioning settings, in which the user can decide if this kind of operations (moving and renaming) should be versioned in the .stversion folder or not.
I personally think that a lot of people would benefit from this, since it would save sometimes hundreds of GB of space and lot of CPU work for something often not necessary (for example if I rename a 500 GB folder, Syncthing will copy 500 GB of the same data in the .stversion folder, practically doubling the space occupied for no reason).

**Comment by acolomb** [comment]
I guess this is a perfect use case for the "external versioning" strategy. You get to decide on the logic of what to keep and where. In Syncthing core, I don't think this will happen.

**Comment by tomasz1986** [comment]
I don't think this is possible with external versioning, as the receiving device can't know whether the file has been renamed, moved, or just deleted. This information would need to come from the sending device.

**Comment by calmh** [comment]
We do rename detection on the receiving side... such as it is; nonetheless, the versioner (external or otherwise) doesn't get told about the reason a file is being versioned.

**Comment by foxxcomm** [comment]
I've run into the same issue over and over again syncing large .VHDX files. Because there is no way to suppress the current behavior I routinely see huge amounts of resources consumed for a small changes in the .VHDX files and/or run out of disk space since the previous .VHDX file is saved before the updates are applied. Even putting the folder on REFS and enabling the duplicate_extents advanced folder option does not help. We need an advanced option to say unsafe updates to a folder are allowed to avoid the overhead. Just have an "Are You Sure?" prompt when enabled explaining the risks.

**Comment by tomasz1986** [comment]
@foxxcomm This sounds like a completely different problem (i.e. to update files in place instead of creating temporary files). This topic is specifically about file versioning.

**Comment by ClemP1** [comment]
I have the same exact problem too!
If it's not possible to add an option in the file versioning settings (as requested by the author of the issue), another possibility could be to add a parameter in the external versioning (in addition to %FOLDER_PATH% and %FILE_PATH%) which contains the information about the reason that file is being versioned (was it modified? renamed? moved?).

**Comment by mherrmann3** [comment]
I'm also interested in reducing the size of `.stversions`, as it typically contains too many files resulting from file renames/moves.

Since those are duplicates of files in the main folder, I wrote me a (Python) script that:
1. scans the synced folder and `.stversions` by creating a hash for each file;
2. identifies files in `.stversions` whose hash is also found in the synced folder ("duplicates");
3. identifies files only *inside* `.stversions` with the same hash ("self-duplicates");
4. reports all duplicates and prompts for their deletion;
5. lastly, removes all (resulting) empty folders.

In my cases with simple file versioning, it reduced the size of `.stversions` by at least 50%.

> Note: Of course, executing it now after years of `.stversions` history, it cannot clean all such duplicates that were ever created. It only cleans duplicates to the *current* state of the synced folder and some (self-)duplicates inside `.stversions` itself - but duplicates to previous file states cannot be identified anymore. But at least that's something. The more often this script is executed, the more duplicates can be identified and purged.
